Google.Api.Gax.Grpc/Rest/PartialDecodingStreamReader.cs

```diff
@@ -0,0 +1,150 @@
+﻿/*
+ * Copyright 2022 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Newtonsoft.Json.Linq;
+
+namespace Google.Api.Gax.Grpc.Rest;
+
+/// <summary>
+/// An IAsyncStreamReader implementation that reads an array of messages
+/// from HTTP stream as they arrive in (partial) JSON chunks. 
+/// </summary>
+/// <typeparam name="TResponse">Type of proto messages in the stream</typeparam>
+internal class PartialDecodingStreamReader<TResponse> : IAsyncStreamReader<TResponse>
+{
+    private readonly Task<StreamReader> _streamReaderTask;
+    private readonly Func<string, TResponse> _responseConverter;
+
+    private readonly Queue<TResponse> _readyResults;
+    private readonly StringBuilder _currentBuffer;
+    
+    private StreamReader _streamReader;
+    private bool _arrayClosed;
+
+    /// <summary>
+    /// Creates the StreamReader
+    /// </summary>
+    /// <param name="streamReaderTask">A stream reader returning partial JSON chunks</param>
+    /// <param name="responseConverter">A function to transform a well-formed JSON object into the proto message.</param>
+    public PartialDecodingStreamReader(Task<StreamReader> streamReaderTask, Func<string, TResponse> responseConverter)
```

Comment:
I think it would be better to use a `TextReader` than a `StreamReader` here. We'd use `StreamReader` for the prod implementation, but logically all we care about is being able to read text data.

That means we can't use `EndOfStream`, but that's okay - I think it makes sense to have a field that knows we've finished reading the data.